sdk/nodejs/cmd/pulumi-language-nodejs/main.go

```diff
@@ -563,115 +544,130 @@ func (host *nodeLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest
 			"It looks like the Pulumi SDK has not been installed. Have you run npm install or yarn install?")
 	}
 
+	// Channel producing the final response we want to issue to our caller. Will get the result of
+	// the actual nodejs process we launch, or any results caused by errors in our server/pipes.
+	responseChannel := make(chan *pulumirpc.RunResponse)
+	defer close(responseChannel)
 	// now, launch the nodejs process and actually run the user code in it.
-	go host.execNodejs(ctx, responseChannel, req, nodeBin, runPath,
-		fmt.Sprintf("127.0.0.1:%d", handle.Port), pipes.directory())
+	go func() {
+		responseChannel <- host.execNodejs(
+			ctx, req, nodeBin, runPath,
+			fmt.Sprintf("127.0.0.1:%d", handle.Port), pipes.directory())
+	}()
 
 	// Wait for one of our launched goroutines to signal that we're done.  This might be our proxy
 	// (in the case of errors), or the launched nodejs completing (either successfully, or with
 	// errors).
-	return <-responseChannel, nil
+	for {
+		select {
+		case err := <-handle.Done:
+			if err != nil {
+				return &pulumirpc.RunResponse{Error: err.Error()}, nil
+			}
+		case err := <-pipesDone:
+			if err != nil {
+				return &pulumirpc.RunResponse{Error: err.Error()}, nil
+			}
+		case response := <-responseChannel:
+			return response, nil
+		}
+	}
 }
 
 // Launch the nodejs process and wait for it to complete.  Report success or any errors using the
 // `responseChannel` arg.
-func (host *nodeLanguageHost) execNodejs(ctx context.Context,
-	responseChannel chan<- *pulumirpc.RunResponse, req *pulumirpc.RunRequest,
```

Comment:
Good call on removing the channel from the function contract here completely.
@dixler was lamenting about something like the problem that caused this panic recently in the context of channels.